File: lab1/bank.cpp
#include "bank.h"
#include <algorithm>
#include <iostream>

Bank::Bank(int num_accounts, int initial_balance_per_account) 
    : initial_total(num_accounts * initial_balance_per_account) {
    
    accounts.resize(num_accounts);
    account_mutexes.resize(num_accounts);
    
    for (int i = 0; i < num_accounts; i++) {
        accounts[i] = std::make_unique<Account>(i, initial_balance_per_account);
        account_mutexes[i] = std::make_unique<std::mutex>();
    }
}

int Bank::get_account_balance(int account_id) {
    if (account_id < 0 || account_id >= static_cast<int>(accounts.size())) {
        return -1;
    }
    std::lock_guard<std::mutex> lock(*account_mutexes[account_id]);
    return accounts[account_id]->get_balance();
}

std::mutex& Bank::get_mutex(int account_id) {
    return *account_mutexes[account_id];
}

bool Bank::transfer(int from_id, int to_id, int amount) {
    if (from_id == to_id) {
        return false;
    }
    
    if (from_id < 0 || from_id >= static_cast<int>(accounts.size()) ||
        to_id < 0 || to_id >= static_cast<int>(accounts.size())) {
        return false;
    }
    
    // Deadlock prevention: always lock in order of account_id
    if (from_id < to_id) {
        std::lock_guard<std::mutex> lock1(*account_mutexes[from_id]);
        std::lock_guard<std::mutex> lock2(*account_mutexes[to_id]);
        
        try {
            accounts[from_id]->withdraw(amount);
            accounts[to_id]->deposit(amount);
            return true;
        } catch (const std::exception&) {
            // Rollback
            accounts[from_id]->deposit(amount);
            return false;
        }
    } else {
        std::lock_guard<std::mutex> lock1(*account_mutexes[to_id]);
        std::lock_guard<std::mutex> lock2(*account_mutexes[from_id]);
        
        try {
            accounts[from_id]->withdraw(amount);
            accounts[to_id]->deposit(amount);
            return true;
        } catch (const std::exception&) {
            // Rollback
            accounts[from_id]->deposit(amount);
            return false;
        }
    }
}

int Bank::get_total_balance() const {
    int total = 0;
    // Lock all accounts to get a consistent snapshot
    std::vector<std::lock_guard<std::mutex>> locks;
    for (auto& mtx : account_mutexes) {
        locks.emplace_back(*mtx);
    }
    
    for (const auto& account : accounts) {
        total += account->get_balance();
    }
    
    return total;
}

bool Bank::consistency_check() const {
    int current_total = get_total_balance();
    return current_total == initial_total;
}

int Bank::get_num_accounts() const {
    return accounts.size();
}

int Bank::get_initial_total() const {
    return initial_total;
}

File: lab1/bank.py
import threading
from account import Account


class Bank:
    def __init__(self, num_accounts, initial_balance_per_account):
        self.initial_total = num_accounts * initial_balance_per_account
        self.accounts = []
        self.locks = []
        
        for i in range(num_accounts):
            self.accounts.append(Account(i, initial_balance_per_account))
            self.locks.append(threading.Lock())
    
    def get_account_balance(self, account_id):
        if account_id < 0 or account_id >= len(self.accounts):
            return -1
        with self.locks[account_id]:
            return self.accounts[account_id].get_balance() 
    
    def transfer(self, from_id, to_id, amount):
        if from_id == to_id:
            return False
        if from_id < 0 or from_id >= len(self.accounts) or to_id < 0 or to_id >= len(self.accounts):
            return False
        
        if from_id < to_id:
            lock1, lock2 = self.locks[from_id], self.locks[to_id]
        else:
            lock1, lock2 = self.locks[to_id], self.locks[from_id]
        
        with lock1, lock2:
            if self.accounts[from_id].get_balance() < amount:
                return False
            self.accounts[from_id].withdraw(amount)
            self.accounts[to_id].deposit(amount)
            return True
    
    def get_total_balance(self):
        for lock in self.locks:
            lock.acquire()
        total = sum(acc.get_balance() for acc in self.accounts)
        for lock in self.locks:
            lock.release()
        return total
    
    def consistency_check(self):
        return self.get_total_balance() == self.initial_total
    
    def get_initial_total(self):
        return self.initial_total
